Source/WebKit/UIProcess/mac/DisplayLink.h

```diff
@@ -27,70 +27,71 @@
-    explicit DisplayLink(WebCore::PlatformDisplayID);
-    ~DisplayLink();
-    
-    void addObserver(IPC::Connection::UniqueID, DisplayLinkObserverID, WebCore::FramesPerSecond);
-    void removeObserver(IPC::Connection::UniqueID, DisplayLinkObserverID);
-    void removeObservers(IPC::Connection::UniqueID);
-
-    void incrementFullSpeedRequestClientCount(IPC::Connection::UniqueID);
-    void decrementFullSpeedRequestClientCount(IPC::Connection::UniqueID);
+    class Client : public CanMakeWeakPtr<Client> {
+    friend class DisplayLink;
+    public:
+        virtual ~Client() = default;
+        
+    private:
+        virtual void displayLinkFired(WebCore::PlatformDisplayID, WebCore::DisplayUpdate, bool wantsFullSpeedUpdates, bool anyObserverWantsCallback) = 0;
+    };
 
-    void setPreferredFramesPerSecond(IPC::Connection::UniqueID, DisplayLinkObserverID, WebCore::FramesPerSecond);
+    DisplayLink(WebCore::PlatformDisplayID);
+    ~DisplayLink();
 
     WebCore::PlatformDisplayID displayID() const { return m_displayID; }
-    
     WebCore::FramesPerSecond nominalFramesPerSecond() const { return m_displayNominalFramesPerSecond; }
 
-    // When responsiveness is critical, we send the IPC to a background queue. Otherwise, we send it to the
-    // main thread to avoid unnecessary thread hopping and save power.
-    static void setShouldSendIPCOnBackgroundQueue(bool value) { shouldSendIPCOnBackgroundQueue = value; }
+    void addObserver(Client&, DisplayLinkObserverID, WebCore::FramesPerSecond);
+    void removeObserver(Client&, DisplayLinkObserverID);
+
+    void removeClient(Client&);
+
+    // FIXME: Maybe callers should just register a DisplayLinkObserverID with the appropriate fps.
+    void incrementFullSpeedRequestClientCount(Client&);
+    void decrementFullSpeedRequestClientCount(Client&);
+
+    void setObserverPreferredFramesPerSecond(Client&, DisplayLinkObserverID, WebCore::FramesPerSecond);
 
 private:
     static CVReturn displayLinkCallback(CVDisplayLinkRef, const CVTimeStamp*, const CVTimeStamp*, CVOptionFlags, CVOptionFlags*, void* data);
     void notifyObserversDisplayWasRefreshed();
 
-    void removeInfoForConnectionIfPossible(IPC::Connection::UniqueID) WTF_REQUIRES_LOCK(m_observersLock);
+    bool removeInfoForClientIfUnused(Client&) WTF_REQUIRES_LOCK(m_clientsLock);
 
     static WebCore::FramesPerSecond nominalFramesPerSecondFromDisplayLink(CVDisplayLinkRef);
 
     struct ObserverInfo {
         DisplayLinkObserverID observerID;
         WebCore::FramesPerSecond preferredFramesPerSecond;
     };
-    
-    struct ConnectionClientInfo {
+
+    struct ClientInfo {
         unsigned fullSpeedUpdatesClientCount { 0 };
         Vector<ObserverInfo> observers;
     };
 
     CVDisplayLinkRef m_displayLink { nullptr };
-    Lock m_observersLock;
-    HashMap<IPC::Connection::UniqueID, ConnectionClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
-    WebCore::PlatformDisplayID m_displayID;
+    Lock m_clientsLock;
```

Comment:
I know it's an existing issue, but:
It is mostly a logical oxymoron to have a weak ptr and a lock together.
Weak ptrs are not thread-safe, and can only be dereferenced on the owner thread.
Then code down the line [1] dereferences the weak ptr in a wrong thread.

Reply:
Replaced the WeakPtr with CheckedPtr.